### ISSUE TITLE
Kucoin price limits

### ISSUE THREAD

**Issue by aloysius-pgast** [ccxt/ccxt#4767]
ATTENTION!!!

MUST READ THIS BEFORE SUBMITTING ISSUES:

https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue

- OS: `Linux`
- Programming Language version: `node v8.12.0`
- CCXT version: `1.18.308`
- Exchange: `Kucoin`
- Method: `fetchMarkets`

Based on [Kucoin's API documentation](https://docs.kucoin.com/#place-a-new-orderl)

> FUNDS
The funds field is optionally used for market orders. When specified it indicates how much of the symbol quote currency to buy or sell. For example, a market buy for BTC-USDT with funds specified as 150.00 will spend 150 USD to buy BTC (including any fees). If the funds field is not specified for a market buy order, size must be specified and Kucoin will use available funds in your account to buy bitcoin.
>
>A market sell order can also specify the funds. If funds is specified, it will limit the sell to the amount of funds specified. You can use funds with sell orders to limit the amount of quote currency funds received.
>
>The funds must be greater than the quoteMinSize for the symbol and no larger than the quoteMaxSize. The size must be specified in quoteIncrement symbol units. Funds indicates the amount of USDT (or quote currency) to buy or sell.

As I understand the documentation, we should have :

* limits.price.max = `quoteMaxSize / baseMinSize` (instead of ` baseMaxSize / quoteMinSize`)
Because buying the *minimum amount* using the *maximum price* per-unit shouldn't be > `quoteMaxSize`
* limits.price.min = `Math.max(quoteMinSize / baseMinSize, quoteIncrement)` instead of `Math.max (baseMinSize / quoteMaxSize, quoteIncrement)`
Because buying the *minimum amount* using the *minimum price* per-unit shouldn't be < `quoteMinSize`

### Examples

For `BTC-USDT`, Kucoin API returns

```
{
    "symbol":"BTC-USDT",
    "quoteMaxSize":"99999999",
    "enableTrading":true,
    "priceIncrement":"0.00000001",
    "baseMaxSize":"10000000000",
    "baseCurrency":"BTC",
    "quoteCurrency":"USDT",
    "market":"SC",
    "quoteIncrement":"0.000001",
    "baseMinSize":"0.00000001",
    "quoteMinSize":"0.000001",
    "name":"BTC-USDT",
    "baseIncrement":"0.00000001"
}
```

#### Using current computation

Using current computation, we will get following values

* limits.price.max = `baseMaxSize / quoteMinSize` = `10000000000 / 0.00000001` = `1e+16`
* limits.price.min = `Math.max (baseMinSize / quoteMaxSize, quoteIncrement)` = `Math.max (0.00000001 / 99999999, 0.000001)` = `Math.max(1.00000001e-16, 0.000001)` = `0.000001`

If I buy the minimum amount of `BTC` (`0.00000001`) for the minimum price per-unit (`0.000001`), I'll end up with a cost of `0.00000001 * 0.000001` = `1e-14` which is `< 0.000001 (quoteMinSize)`

If I buy the minimum amount of `BTC` (`0.00000001`) for the maximum price per-unit (`1e+16`), I'll end up with a cost of `0.00000001 * 1e+16` = `100000000` which is `> 99999999 (quoteMaxSize)`

#### Using proposed computation

* limits.price.max = `quoteMaxSize / baseMinSize`  = `99999999 / 0.00000001` = `9.9999999e+15`
* limits.price.min = `Math.max(quoteMinSize / baseMinSize, quoteIncrement)` = `Math.max (0.000001 / 0.00000001, 0.000001)` = `Math.max(100, 0.000001)` = `100`

If I buy the minimum amount of `BTC` (`0.00000001`) for the minimum price per-unit (`100`), I'll end up with a cost of `0.00000001 * 100` = `1e-6` which is `= 0.000001 (quoteMinSize)`

If I buy the minimum amount of `BTC` (`0.00000001`) for the maximum price per-unit (`9.9999999e+15`), I'll end up with a cost of `0.00000001 * 9.9999999e+15` = `99999999` which is `= 99999999 (quoteMaxSize)`

**Comment by kroitor** [comment]
@aloysius-pgast your comments on this issue are very welcome: https://github.com/ccxt/ccxt/issues/4858 ;)

**Comment by aloysius-pgast** [comment]
Hi @kroitor. I agree it would make more sense to not use `Math.max` and use `quoteIncrement` directly. After re-reading Kucoin's documentation, I see that it says
> The baseMinSize and baseMaxSize fields define the min and max order size. The priceIncrement field specifies the min order price as well as the price increment.This also applies to quote currency.

There is no way to satisfy all conditions (rate, cost & quantity) at the same time anyway. With the fix for #4858, we would have the following :

* *min(amount)* = `baseMinSize` = `0.01`
* *min(price)* = `quoteIncrement` = `0.00001`
* *min(cost)* = `quoteMinSize` = `0.05`

If we buy the minimum amount for the minimum price, cost will be `0.01 * 0.00001 = 0.0000001` which won't be valid because of *min(cost)*

But it's probably better to let the end-user validate that *min(cost)* condition is satisfied by himself and stick to `quoteIncrement`, although it would be good to get a confirmation from Kucoin on that. Since ccxt is the official sdk for Kucoin v2, do you think it would be possible to contact them to clarify whether or not min rate should indeed be the same as `quoteIncrement` ?

**Comment by aloysius-pgast** [comment]
One more thing ;) Base on previous doc extract
> The baseMinSize and baseMaxSize fields define the min and max order size. The priceIncrement field specifies the min order price as well as the price increment.This also applies to quote currency.

My understanding was that `priceIncrement` is for the order price (ie: `cost` in *ccxt*), not for the rate (ie: `price` in *ccxt*). So I think we should use `quoteIncrement` for `price.min`, not `priceIncrement`.

I might be wrong though (Kucoin's doc is not clear enough imho and only them could provide a better understanding). Anyway, using `quoteIncrement` or `priceIncrement` should be better than using `Math.max`. 

Also, I think that if I spend one more minute trying to interpret Kucoin's doc, I'm gonna go insane ;)
